### PR TITLE
Intgrtns 237 add per domain syncs button in whmcs admin area

### PR DESCRIPTION
**Pull request by DaniNC**

### CODE REVIEW

File: modules/registrars/openprovider/src/DomainSyncAutoRenew.php [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+    private $idn;
+    
+    /**
+     * Init the class
+     *
+     * @return void
+     **/
+    public function __construct(ApiHelper $apiHelper, \idna_convert $idn, $limit = false)
+    {
+        var_dump('DomainSyncAutoRenew');die;
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/src/DomainSyncAutoRenew.php:97]: remove this

File: modules/registrars/openprovider/Controllers/System/DomainSyncAutoRenewController.php [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+     * Synchronise the transfer status.
+     *
+     * @param $params
+     * @return array
+     */
+    public function sync($params)
+    {
+        $this->domain = $this->domain->find($params['domainid']);
+        // Check if the native synchronisation feature
+        if(Configuration::getOrDefault('syncUseNativeWHMCS', false) == false) {
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/Controllers/System/DomainSyncAutoRenewController.php:60]: better to use ===

File: modules/registrars/openprovider/src/DomainSyncAutoRenew.php [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+                ];
+
+                $this->printDebug('PROCESSING DOMAIN STATUS CHANGE ' . $this->objectDomain->domain .' ( ' . $this->objectDomain->status . ' => ' . $op_domain_status . ')');
+
+                $this->update_executed_logs[] = ['activity' => 'update_domain_status', 'data' => $activity_data];
+
+            }
+
+        }
+        else
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/src/DomainSyncAutoRenew.php:436]: I can see most of the places, you added curly braces in new lines, for consitions better to use same like you did try catch

File: modules/registrars/openprovider/src/DomainSyncAutoRenew.php [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+
+    /**
+     * Process the Domain status setting
+     *
+     * @param  string $status Set the domain status
+     * @return void
+     **/
+    private function process_domain_status($status = null)
+    {
+        if($status == 'Cancelled' || $status == 'Expired' || $status == 'Transferred Away')
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/src/DomainSyncAutoRenew.php:375]: try to use ===

File: modules/registrars/openprovider/src/DomainSyncAutoRenew.php [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+
+        // Sync the expiry date
+        $whmcs_expiry_date = new Carbon($this->objectDomain->expirydate);
+
+        // Check if we have a valid timestamp.
+        if($whmcs_expiry_date->getTimestamp() > 0)
+        {
+            $expiry_date_result = General::compare_dates($whmcs_expiry_date->toDateString(), $this->op_domain['renewalDate'], '0', 'Y-m-d H:i:s');
+        }
+        else
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/src/DomainSyncAutoRenew.php:277]: this too, curly in same line, more readble

File: modules/registrars/openprovider/src/DomainSyncAutoRenew.php [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+     */
+    private function check_if_domain_deleted(): bool
+    {
+        try {
+            $this->op_domain = $this->apiHelper->getDomain($this->op_domain_obj, [
+                'isDeleted' => true,
+            ]);
+
+            $this->process_domain_status('Cancelled');
+            return true;
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/src/DomainSyncAutoRenew.php:513]: put a new line before return

File: modules/registrars/openprovider/src/DomainSyncAutoRenew.php [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+            if($this->objectDomain->status != $op_domain_status)
+            {
+                if($this->objectDomain->status == 'Pending Transfer' && $op_domain_status == 'Active')
+                {
+                    if($this->objectDomain->check_renew_domain_setting_upon_completed_transfer() == true)
+                    {
+                        $this->apiHelper->renewDomain($this->op_domain['id'], $this->objectDomain->registrationperiod);
+
+                        // Fetch updated information
+                        $this->op_domain             =    $this->apiHelper->getDomain($this->op_domain_obj);
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/src/DomainSyncAutoRenew.php:410]: why this so spaces

File: modules/registrars/openprovider/src/DomainSyncAutoRenew.php [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+        }
+
+        $op_domain_status = api_domain::convertOpStatusToWhmcs($this->op_domain['status']);
+
+        if($op_domain_status != false)
+        {
+            // Check if the status matches
+            if($this->objectDomain->status != $op_domain_status)
+            {
+                if($this->objectDomain->status == 'Pending Transfer' && $op_domain_status == 'Active')
```
Comment (sampikaipakan) [inline modules/registrars/openprovider/src/DomainSyncAutoRenew.php:403]: too much nested seems

**Review by sampikaipakan — CHANGES_REQUESTED** [review]

**Review by sapillai — APPROVED** [review]
Tested the branch with latest changes, it is working as expected. Expiration date and status were synced correctly to WHMCS.

**Review by chhatbarhiren — APPROVED** [review]
I checked the code, didn't make sense to me but I trust Sharmi and @sapillai's review and his testing and hence, approving for good. Thank you.

**Pull request merged**